src/python/pants/core/goals/test.py

```diff
@@ -481,28 +625,49 @@ class TestExtraEnvVarsField(StringSequenceField, metaclass=ABCMeta):
     )
 
 
+@rule_helper
+async def _get_partitions_by_request_type(
```

Comment:
Why is the return type dict's value a `list[Partitions]` and not just a `Partitions`? AFAICT from the implementation it will always contain exactly one `Partitions`?

Reply:
I ask because the multiple levels of plurality are a mite confusing...

Reply:
If so then right now those concepts seem mingled. E.g., `partition_per_input()` is on Partition, not SubPartition. 

We had discussed the terminology on Slack I believe? Should that be introduced here?

I do think it makes sense to have separate rules for the two different concepts.

Reply:
The terminology discussion on Slack was mainly around what field names to expose to end users, which (purposefully) isn't touched in this PR - the intent here is to get the generic plumbing in place.

One of my take-aways from this PR is that it's hard to effectively design/review generic plumbing without a concrete implementation to look at 😅

Reply:
Gotcha. The terminology, and functionality, for plugin authors and end users should probably be consistent though? 

Right now it's not clear to me where the logic for sub-batching-by-size would go, for example.